src/lib/actions.tsx

```diff
@@ -0,0 +1,161 @@
+import { Message } from 'ai';
+import {
+  createAI,
+  createStreamableUI,
+  createStreamableValue,
+  getMutableAIState,
+  render,
+} from 'ai/rsc';
+import { nanoid } from 'nanoid';
+import { OpenAI } from 'openai';
+
+import { UserMessage, WonkMessage } from '@/components/chat/chatMessage';
+import {
+  getEmbeddings,
+  getSearchResults,
+  transformSearchResults,
+  getSystemMessage,
+} from '@/services/chatService';
+
+import { AIState, UIState } from './types';
+
+const openai = new OpenAI({
+  apiKey: process.env.OPENAI_API_KEY,
+});
+
+const llmModel = process.env.OPENAI_LLM_MODEL ?? 'gpt-3.5-turbo';
+
+async function submitUserMessage(userInput: string) {
+  'use server';
+
+  const userMsgId = nanoid();
+  const wonkMsgId = nanoid();
+
+  // before we actually do anything, stream loading UI (for the chat window)
+  // would be better to add this on the client side so it is immediate
+  // but putting it here for now
+  const chatWindowUI = createStreamableUI(
```

Comment:
this is from ai lib, is streamableUI specific to the AI library or is this a common RSC pattern we could do with other libs?

Reply:
@srkirkland createStramableUI is specific to ai/rsc, but it's just using suspense and promises underneath. what you get from SSR is the ability to stream components in chunks–the chunks being determined by being async components or wherever you specify a suspense boundary. in next these are automatically set at the page level, which is why you can do `loading.js` and `error.js`  

https://github.com/vercel/ai/blob/e3206ff19b850d40dd9cbbe53d9b7a93418bf622/packages/core/rsc/streamable.tsx#L24
https://github.com/vercel/ai/blob/e3206ff19b850d40dd9cbbe53d9b7a93418bf622/packages/core/rsc/utils.tsx#L52

suspense works by throwing promises and using the status to determine when to switch states for rendering. like: 
```js
  if (fetchRequest.status === 'pending') {
    throw fetchRequest;
  }
  ```
🤯 
so while it's pending, it uses whatever is supplied as the fallback. so looks like they're just not resolving that outer promise until you call like `ui.done()`. and within that, to update the actual react node before you call `.done()`, they are also leveraging suspense boundaries and promisifying components. 

so to update a node: 
```js
export function createStreamableUI(initialValue?: React.ReactNode) {
 // create outer suspense boundary and promise
  let { row, resolve, reject } = createSuspensedChunk(initialValue);
  ...
  return {
  ...
    append(value: React.ReactNode) {
      assertStream('.append()');

      const resolvable = createResolvablePromise();
      currentValue = value;

      // resolve current inner promise, but make a new one inside a suspense boundary 
      // which will display our currentValue until it is resolved (via 
      resolve({ value, done: false, append: true, next: resolvable.promise }); 
      resolve = resolvable.resolve;
      reject = resolvable.reject;
      ...

 ```
 
 and that resolve in append moves the suspense boundary around a new inner promise
 ```js
     ...
     if (chunk.append) {
      return (
        <>
          // what we had before appending, if !chunk.append this is just discarded 
          {currentNode} 
          // make a new suspense boundary but make the fallback the ReactNode 
          // (this is the wildest part to me) 
          // so the actual ReactNode we are rendering is actually rendered as the fallback
          //  BEFORE the promise is resolved
          //  and when we are resolving the promise that represents the currentNode
          // we create the next suspended node, suspended on the next promise
          <Suspense fallback={chunk.value}> 
            <Row currentNode={chunk.value} nextPromise={chunk.next} />
          </Suspense>
        </>
      );
      ...
    }
    ```
 
 so tl;dr yes we can do that with just RSC (though i imagine there are/will be libraries to abstract this)